File: frontend/src/lib/types/Ticket.ts
// Relation between a user and an event
export interface Ticket {
  ticketId: string;
  userId: string;
  eventId: string;
  role: USER_ROLE;
  accessCode?: string; //for virtual event
  virtualLink?: string;
  qrCode?: string; //for inPerson event
  registrationDate: Date;
}

export enum USER_ROLE {
  ORGANIZER = "organizer",
  ATTENDEE = "attendee",
  SPEAKER = "speaker",
  SPONSOR = "sponsor",
  STAKEHOLDER = "stakeholder",
  EVENT_ADMIN = "eventAdmin",
}

export interface BonusTicket {
  accessCode: string;
}

File: backend/models/ticket.py
from sqlmodel import Field, SQLModel
from enum import Enum
import datetime as dt


class UserRole(str, Enum):
    ORGANIZER = "organizer",
    ATTENDEE = "attendee",
    SPEAKER = "speaker",
    SPONSOR = "sponsor",
    STAKEHOLDER = "stakeholder",
    EVENT_ADMIN = "eventAdmin",


class TicketBase(SQLModel):
    user_id: int
    event_id: int


class Ticket(TicketBase):
    role: UserRole = Field(default=UserRole.ATTENDEE)
    access_code: str | None = Field(default=None)
    virtual_link: str | None = Field(default=None)
    qr_code: str | None = Field(default=None)
    registration_date: dt.datetime


class DbTicket(Ticket, table=True):
    __tablename__ = "tickets"
    id: int | None = Field(default=None, primary_key=True)


class TicketPublic(Ticket):
    id: int
